### PR TITLE
Actinn featurizer

### PR DESCRIPTION
**Pull request by Harindhar10**
## Description

Draft PR to merge the featurizer for ACTINN. 
 
Works with DeepChem's CSVLoader class (.h5 file is converted to .csv file)

What does the featurizer do?
It normalizes scRNA-seq data and filters genes based on expression

Both the train and test sets have to be processed together, but only the train set processing is implemented so far. 
Adding the test set to it requires a new dataloader

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/__init__.py [hunk 1]
```diff
@@ -106,3 +106,6 @@
 
 # support classes
 from deepchem.feat.molecule_featurizers import GraphMatrix
+
+# single cell rna sequencing data featurizers
+from deepchem.feat.scRNAseq import ACTINNFeaturizer
```
Comment (sudo-rsingh) [inline deepchem/feat/__init__.py]: Pls leave an empty line at the end of file

File: deepchem/feat/scRNAseq.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
+from deepchem.feat import Featurizer
+import numpy as np
+import pandas as pd
+import logging
+from typing import Any, Iterable
+from deepchem.data import NumpyDataset
+from typing import Tuple
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]: Can you please sort them properly, like 
```suggestion
import logging
from typing import Any, Iterable, Tuple

import numpy as np
import pandas as pd

from deepchem.data import NumpyDataset
from deepchem.feat import Featurizer
```

File: deepchem/feat/scRNAseq.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+logger = logging.getLogger(__name__)
+
+
+class ACTINNFeaturizer(Featurizer):
+    """
+    This class is the implementation of all transformations performed to
+    scRNA-seq (Single-cell RNA sequencing) data for cell type identification
+    using the model ACTINN.
+
+    Example:
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]:
```suggestion
    Example
```

File: deepchem/feat/scRNAseq.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    <NumpyDataset X.shape: (1000, 13389), y.shape: (1000,), w.shape: (1000,), task_names: [0]>
+    >>> test_dataset = featurizer.featurize(train_set,labels=labels[1],mode='test')
+    >>> test_dataset
+    <NumpyDataset X.shape: (1000, 13389), y.shape: (1000, 1), w.shape: (1000, 1), task_names: [0]>
+
+
+    References:
+    -----------
+    [1] https://academic.oup.com/bioinformatics/article/36/2/533/5540320
+
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]: pls don't give random empty line

File: deepchem/feat/scRNAseq.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    >>> train_labels = pd.read_csv('os.path.join(dataset,"train_label.txt")',sep='\t',header=None)
+
+    >>> featurizer = dc.feat.ACTINNFeaturizer()
+
+    >>> train_dataset = featurizer.featurize(train_set,labels[1])
+    >>> train_dataset
+    <NumpyDataset X.shape: (1000, 13389), y.shape: (1000,), w.shape: (1000,), task_names: [0]>
+    >>> test_dataset = featurizer.featurize(train_set,labels=labels[1],mode='test')
+    >>> test_dataset
+    <NumpyDataset X.shape: (1000, 13389), y.shape: (1000, 1), w.shape: (1000, 1), task_names: [0]>
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]: Pls give space after `,`. that makes things neat.

File: deepchem/feat/scRNAseq.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    def convert_type2label(self, types, type_to_label_dict):
+        """
+        Convert types to labels
+
+        Parameters:
+        -----------
+            types: List
+                list of cell types
+            type_to_label dictionary: Dict
+                dictionary of cell types mapped to integer labels
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]: Pls fix indentations

File: deepchem/feat/scRNAseq.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        **kwargs :
+            Additional arguments passed to the internal `_featurize` function.
+
+        Returns
+        -------
+        NumpyDataset
+            A dataset object containing the transformed features and corresponding labels.
+        """
+        self.labels = labels
+        print('self.labels', self.labels)
```
Comment (sudo-rsingh) [inline deepchem/feat/scRNAseq.py]: pls remove print statement, as it is a production code

**Review by sudo-rsingh — CHANGES_REQUESTED** [review]

File: deepchem/feat/scRNAseq.py (hunk 2 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: done

File: deepchem/feat/scRNAseq.py (hunk 3 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: done

File: deepchem/feat/scRNAseq.py (hunk 5 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: done

File: deepchem/feat/__init__.py (hunk 1 above)
Comment (Harindhar10) [reply deepchem/feat/__init__.py]: flake8 shows a warning, and yapf removes the blank line

File: deepchem/feat/scRNAseq.py (hunk 4 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: removed

File: deepchem/feat/scRNAseq.py (hunk 6 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: fixed

File: deepchem/feat/scRNAseq.py (hunk 7 above)
Comment (Harindhar10) [reply deepchem/feat/scRNAseq.py]: removed

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+logger = logging.getLogger(__name__)
+
+
+class ACTINNFeaturizer(Featurizer):
+    """
+    This class is the implementation of all transformations performed to
+    scRNA-seq (Single-cell RNA sequencing) data for cell type identification
+    using the model ACTINN.
+
+    References
```
Comment (sudo-rsingh) [inline deepchem/feat/sc_rna_seq_featurizers.py]: pls keep the reference at the end of the doc string

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+    Examples
+    --------
+    >>> import deepchem as dc
+    >>> from deepchem.data import NumpyDataset
+    >>> import pandas as pd
+    >>> import os
+    >>> featurizer = dc.feat.ACTINNFeaturizer()
+    >>> dataset = os.path.join(os.path.dirname(__file__), "tests",
+    ...                            "data", "sc_rna_seq_data")
```
Comment (sudo-rsingh) [inline deepchem/feat/sc_rna_seq_featurizers.py]:
```suggestion
    ...                        "data", "sc_rna_seq_data")
```

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+            Parameters
+            ----------
+            datapoints: Iterable[Any]
+                Pandas Dataframe containing the dataset
+
+            Returns
+            -------
+            np.ndarray
+                A numpy array containing a featurized representation of `datapoints`.
```
Comment (sudo-rsingh) [inline deepchem/feat/sc_rna_seq_featurizers.py]: pls fix indentation mistake

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        labels = list()
+        for type in types:
+            labels.append(type_to_label_dict[type])
+        return labels
+
+    def featurize(self,
+                  datapoints: Iterable[Any],
+                  log_every_n: int = 1000,
+                  **kwargs) -> np.ndarray:
+        """  Converts raw scRNA-seq data into model-ready features and labels.
```
Comment (sudo-rsingh) [inline deepchem/feat/sc_rna_seq_featurizers.py:147]:
```suggestion
        """Converts raw scRNA-seq data into model-ready features and labels.
```

File: deepchem/feat/sc_rna_seq_featurizers.py (hunk 8 above)
Comment (Harindhar10) [reply deepchem/feat/sc_rna_seq_featurizers.py]: done

File: deepchem/feat/sc_rna_seq_featurizers.py (hunk 10 above)
Comment (Harindhar10) [reply deepchem/feat/sc_rna_seq_featurizers.py]: done

**Review by sudo-rsingh — APPROVED** [review]
LGTM

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
+from deepchem.feat import Featurizer
+import numpy as np
+import pandas as pd
+import logging
+from typing import Any, Iterable, Tuple, List
+
+logger = logging.getLogger(__name__)
+
+
+class ACTINNFeaturizer(Featurizer):
```
Comment (rbharath) [inline deepchem/feat/sc_rna_seq_featurizers.py:10]: Please add to the rst docs

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    """
+    This class is the implementation of all transformations performed to
```
Comment (rbharath) [inline deepchem/feat/sc_rna_seq_featurizers.py:12]: More detail; which transformations are these? What do they do? Please add a citation for the scientific context

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        scaled_set = self.normalise(dataset)
+        mask = self.filter_genes()
+        scaled_set = scaled_set[mask, :]
+
+        # gens x cells --> cells x gens
+        dataset = np.transpose(scaled_set)
+
+        return dataset
+
+    def normalise(self, dataset_df: pd.DataFrame, **kwargs) -> np.ndarray:
```
Comment (rbharath) [inline deepchem/feat/sc_rna_seq_featurizers.py:209]: nit, but normalize (we use US and not british spelling)

File: deepchem/feat/sc_rna_seq_featurizers.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+    ...                        "data", "sc_rna_seq_data")
+    >>> # train_set.shape = (genes x cells)
+    >>> train_set = pd.read_hdf(os.path.join(dataset, "scRNAseq_sample_1.h5"))
+    >>> train_set.shape
+    (1500, 1000)
+    >>> # test_set.shape = (genes x cells)
+    >>> test_set = pd.read_hdf(os.path.join(dataset, "scRNAseq_sample_2.h5"))
+    >>> test_set.shape
+    (1200, 1000)
+    >>> train_set,test_set = featurizer.find_common_genes(train_set, test_set)
```
Comment (rbharath) [inline deepchem/feat/sc_rna_seq_featurizers.py:33]: Splitting data into train/test is the work of a splitter in DeepChem
